Feature/branches coverage support

som4ik opened the pull request:
Add branch coverage measurement statistics (ruby ~> 2.5) 
In this PR:
- Add configuration needed to trigger this new functionality.
```ruby
SimpleCov.start do
  use_branchable_report true
end
```
  If config is set for ruby versions which are < 2.5, It will shows notify message and produce only lines coverage report.
- Unite the results no matter what coverage version used to be there. Now always the coverage result is a hash with keys represents the type of coverage and values represents the coverage result for each type ex: `{:lines => ..., :branches => ..}`.
- Add `Branch` class with related support methods to `SourceFile`
- Replace `RawCoverage` with `RunResultsCombiner`, which breaks logic of combining different results into file, line and branch combining level, this new structure adds more readability and makes it open for modification (ex: adding methods coverage or any other coverage in future)
- Remove useless files from the coverage result before any operation on it, as `Coverage.start` take in case all loaded files including required gems, which files are not inside the project directory.
- On simplecov-html side we will have this visual statistics:
![branch1](https://user-images.githubusercontent.com/7699543/43396656-e273be44-940a-11e8-99b3-053929752e3d.png)
 `[0, '-'] => [passing count, branch sign]`
Condition with missed branch will be marked similar to line 70.

Benefits:
- Coverage report gives more information about conditions especially the inline conditions similar to the case specified above.

fwininger commented:
Could we consider this PR ? It's very usefull for our project.

klyonrad commented:
Hey @som4ik , I got very excited after your comments in the issue and tried out your branch in my current work project.

Sadly I encountered an exception after the test suite ran. Maybe the simple CallTrace can help you already; next week I might have a deeper insight with debugging. I hope that is not one specific file that caused this exception 🤣 

Our docker base image is phusion passenger ruby25


```
..../usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov/branches_per_file.rb:17:in `start': undefined method `[]' for nil:NilClass (NoMethodError)
	from /usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov/run_file_coverage.rb:26:in `call'
	from /usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov/run_file_coverage.rb:9:in `start'
	from /usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:265:in `block in add_not_loaded_files'
	from /usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:263:in `each'
	from /usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:263:in `add_not_loaded_files'
	from /usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:299:in `result_with_not_loaded_files'
	from /usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:310:in `process_coverage_result'
	from /usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:70:in `result'
	from /usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov/configuration.rb:191:in `block in at_exit'
	from /usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:182:in `run_exit_tasks!'
	from /usr/local/rvm/gems/ruby-2.5.1/bundler/gems/simplecov-52827cb42b93/lib/simplecov/defaults.rb:27:in `block in <top (required)>'


Finished in 1394.358260s, 2.3889 runs/s, 4.2851 assertions/s.
3331 runs, 5975 assertions, 0 failures, 0 errors, 0 skips
```

klyonrad commented on lib/simplecov/branches_per_file.rb:
```diff
@@ -0,0 +1,20 @@
+    def self.start(source_file_path)
+      return {} unless SimpleCov.branchable_report
+      Coverage.start(:all)
+      require source_file_path
```
Are you sure that the `require` is stable? I am receiving a loooot of `LoadError` errors when this is executed

som4ik replied:
Well, thank you for being interested in this PR :), l had a look on your comments: 
1 - In line 16 `require` is called on the absolute file path so it's safe for the file to be required. `require` can through and LoadError if file/module target is not located in global load path constant.  
2 - About the non-nested style of defining modules, it should through another type of errors `NameError`.
3 - The logs that provided me with is `branches_per_file.rb:17:in `start': undefined method `[]'` the line `17`. This type of error is a result of totally empty ruby file which it's coverage report return an empty hash `{}`, **I will fix it**.
4 - Are you using `track_file` configuration on simplecov? could you please try to require the file that you have and with absolute path of it in the irb console? (This will help me to better understand the issue)
5-  `BranchesPerFile` concern the files that are missing any coverage report (not covered by applied tests), so it will return just the branches with zero coverage result. 
As a part of the solution on this case, is to write something similar to `LinesClassifier` that will parse the lines and give the report of branches, but this will make this PR very complicated to review. Better to have a simple solution that will return empty hash for now, and make a new PR with the new functionality of text parsing after this one will be accepted.
I appreciate your interest in this PR, hope our collaboration will make SimpleCov have this feature soon.

klyonrad replied:
1: 
> require is called on the absolute file path so it's safe

I do not quite understand what the file path has to do with safety of the file. My point was that with this change there is an unexpected eager loading on the execution of the test suite

3: This error appears also in a different case: execution of only a specified test or a specified directory

Example:
```
rspec spec/models

Randomized with seed 24924
..................................................

Finished in 1.89 seconds (files took 2.47 seconds to load)
50 examples, 0 failures

Randomized with seed 24924

Traceback (most recent call last):
	11: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/defaults.rb:27:in `block in <top (required)>'
	10: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:182:in `run_exit_tasks!'
	 9: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/configuration.rb:191:in `block in at_exit'
	 8: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:70:in `result'
	 7: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:310:in `process_coverage_result'
	 6: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:299:in `result_with_not_loaded_files'
	 5: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:263:in `add_not_loaded_files'
	 4: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:263:in `each'
	 3: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:265:in `block in add_not_loaded_files'
	 2: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/run_file_coverage.rb:9:in `start'
	 1: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/run_file_coverage.rb:26:in `call'
/Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/branches_per_file.rb:17:in `start': undefined method `[]' for nil:NilClass (NoMethodError)

```

4: No, am not using `track_file`

Sorry, I don't have time for this at the day-to-day work anymore; so I cannot get more into the details what is wrong there.

However, I made some of the problematic examples that came to my mind in my (much simple) pet project.

Have a look at https://github.com/klyonrad/lazybar/commits/simplecov-branch-cov

the full `rspec` run actually works fine with the edge cases that I wrote but the the execution of a directory or a single file breaks - with a different error then what I copied in my previous example:

```
rspec spec/models/cocktail_recipe_spec.rb

Randomized with seed 20614
..................

Top 2 slowest examples (1.15 seconds, 55.6% of total time):
  CocktailRecipe validations is not valid with only one ingredient
    1.05 seconds ./spec/models/cocktail_recipe_spec.rb:21
  CocktailRecipe methods about liking can show the users that like it
    0.09874 seconds ./spec/models/cocktail_recipe_spec.rb:70

Finished in 2.07 seconds (files took 3.19 seconds to load)
18 examples, 0 failures

Randomized with seed 20614

Traceback (most recent call last):
	13: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/defaults.rb:27:in `block in <top (required)>'
	12: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:182:in `run_exit_tasks!'
	11: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/configuration.rb:191:in `block in at_exit'
	10: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:70:in `result'
	 9: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:310:in `process_coverage_result'
	 8: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:299:in `result_with_not_loaded_files'
	 7: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:263:in `add_not_loaded_files'
	 6: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:263:in `each'
	 5: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:265:in `block in add_not_loaded_files'
	 4: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/run_file_coverage.rb:9:in `start'
	 3: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/run_file_coverage.rb:26:in `call'
	 2: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/branches_per_file.rb:16:in `start'
	 1: from /Users/luka/development/lazy_bar/app/models/never_tested.rb:1:in `<top (required)>'
/Users/luka/development/lazy_bar/app/models/never_tested.rb:2:in `<class:NeverTested>': uninitialized constant NeverTested::NotExistingNamespace (NameError)
```

As I said: even though it would be a good thing to catch all these type of stupid programmer errors; I think an autoloading of all code is a bit unexpected behaviour

som4ik replied:
@PragTob  see my comment above if you still not agree I will remove this class, and add it to my TODO list for next PR :)

stepozer replied:
I suggest remove this class, because we tested your PR and found that took 10-15 seconds for our big legacy project... After remove this class everything working fast.

som4ik replied:
@stepozer  Yep, that's what exactly I will do. We already agreed to do it at the comments above.

Marc3001 commented:
Is anyone available to review this PR ? We were really grateful to see this feature in simplecov on our ruby2.5 projects ;)

jeremyevans commented:
I've wanted to use support branch coverage for a while in my projects, so I tried out this pull request (and the corresponding one in simplecov-html).  Here is what I found, and they may well be issues with branch support in stdlib coverage and not issues with this pull request:

* In general, this works well and adds very useful information on branch coverage.  I found numerous places with missing branch coverage in projects I maintain that have 100% line coverage.
* Some lines that should be covered and have line coverage are missing branch coverage.  For example a line with `%w'foo bar'.each do |x|` or `self::Constant.class_eval do` will be marked as line covered but not branch covered (`[0, '-']`).  Maybe lines that have line coverage but only a single branch with no branch coverage should be considered covered? 
* Some empty lines are reported with missed branch coverage with `[0, "+"]`.  Maybe if the line isn't hit at all, we shouldn't report branch coverage on it?  Lines marked this way are not handled properly in the html view, since the end up with a height of only a couple pixels.
* Some starting comment lines with `# :nocov:` are marked as skipped and still marked red with missing branch coverage with `[0, '-']`.  Additionally, some other lines inside nocov blocks are marked as skipped with missing branch coverage (`when`/`else` in `case`).  If a line is skipped, branch coverage should be skipped for that line.

I would definitely like to see branch support in simplecov.  Hopefully the issues can be addressed and this can be merged.

PragTob commented:
Hi everyone!

Thanks @som4ik for all the work on this, everyone for their encouragements, testing this out and reporting problems :tada: 

It seems like some bugs still might need to be squashed here before we can release it. I'm slowly getting back to trying to maintain simplecov and this is of course a huge piece of work. I'll aim to review and determine what might/might not need doing before we can land it and maybe release it as an "alpha" feature or what not.

Please bear with me for a little while longer, this is a huge PR.

@som4ik: Which one is the preferred PR to review? This here or #692 ? I'm a bit confused because #692 says "the first time I developed this" but points to this PR which was made after #692 ? Which one do you prefer to get in/which one should I review/work on?

Also bunny for your efforts:

![IMG_20180805_081900](https://user-images.githubusercontent.com/606517/60168574-63e94080-9805-11e9-8a15-33c8c51a654a.jpg)


Thanks,
Tobi

som4ik commented:
@PragTob  Thanks, I accepted the cute bunny :smiley: 
Frankly speaking, I'd rather you had a look at the current PR #694 , I know it's quite big but it's structured way better and will support us in feature to expand and improve the functionality.
Looking forward to your comments.

PragTob commented:
I see, thanks for the look. Aim is to take a look in the coming days, maybe even today. :)

PragTob commented on lib/simplecov/supports/source_file_support.rb:
```diff
@@ -0,0 +1,221 @@
+      #
+      # @return [Array]
+      #
+      def branches_collection(given_branches, root_id = nil)
```
Recursive methods in Ruby have me scared that we might blow the stack :scream: (due to no tail call optimization)

Probably not something for now, but something we might look at if we find a non recursive way to do this. It does sound like a quite naturally recursive problem though, I should look how big these usually get. If this is just called for a branch of one method we should be fine.

PragTob commented on lib/simplecov/supports/source_file_support.rb:
```diff
@@ -0,0 +1,221 @@
+      # @return [Array]
+      #
+      def branches_collection(given_branches, root_id = nil)
+        @branches_collection ||= []
```
we never use `@branches_collection` again, right? Could just be a local variable that we pass along then, right?

PragTob replied:
(renamed though due to method name clash

PragTob commented on lib/simplecov/supports/source_file_support.rb:
```diff
@@ -0,0 +1,221 @@
+          if value.is_a?(Integer)
+            branch.coverage = value
+          else
+            branches_collection(value, branch.id)
```
Maybe we're not doing this here, but I'd expect the "mother branch" to have the summed up value of its children... maybe that doesn't make sense though and we need it nowhere :)

PragTob commented on lib/simplecov/supports/source_file_support.rb:
```diff
@@ -0,0 +1,221 @@
+      #
+      def covered_branches
+        @coverd_branches ||= root_branches.flat_map do |root_branch|
+          root_branch.sub_branches(branches).select(&:covered?)
```
it's really hard to determine where `branches` here comes from. I suspect from the including class

PragTob commented on lib/simplecov/supports/source_file_support.rb:
```diff
@@ -0,0 +1,221 @@
+      #
+      # @param [Integer] line_number
+      #
+      # @return [String] ex: "[1, '+'],[2, '-']" two times on negative branch and non on the positive
```
shouldn't `[1, '+']` mean one hit on the positive branch?

PragTob commented on spec/run_results_combiner_spec.rb:
```diff
@@ -0,0 +1,107 @@
+  describe SimpleCov::RunResultsCombiner do
+    describe "with two faked coverage resultsets" do
+      before do
+        @resultset1 = {
```
Instead of using `before` and `@resultset` I think `let(:resultset1)` etc. should do the job - what do you think? :)

PragTob commented on spec/run_results_combiner_spec.rb:
```diff
@@ -0,0 +1,107 @@
+            :branches => {[:if, 3, 8, 6, 8, 36] => {[:then, 4, 8, 6, 8, 12] => 47, [:else, 5, 8, 6, 8, 36] => 24}},
+          },
+          source_fixture("app/controllers/sample_controller.rb") => {:lines => [nil, 1, 1, 1, nil, nil, 1, 0, nil, nil]},
+          source_fixture("resultset1.rb") => {:lines => [1, 1, 1, 1]},
```
is it common that there will be a `:lines` key but no `:branches` key? I'd ahve expected both of them to always be present

PragTob commented on spec/run_results_combiner_spec.rb:
```diff
@@ -0,0 +1,107 @@
+        end
+
+        it "has proper results for three.rb" do
+          expect(subject[source_fixture("three.rb")][:lines]).to eq([nil, 3, 7])
```
I'll just believe you that all of these are correct :angel:

PragTob commented on spec/source_file_branch_spec.rb:
```diff
@@ -0,0 +1,151 @@
+
+if SimpleCov.usable?
+  describe SimpleCov::SourceFile::Branch do
+    let(:results) do
```
this `results` confuses me a bit, I think it's only used in the other `let`s to do `result[index]` - coudln't we just inline them then? Maybe I just don't understand its purpose :sweat_smile:

som4ik replied:
I will modify it.

PragTob commented on spec/source_file_branch_spec.rb:
```diff
@@ -0,0 +1,151 @@
+      end
+
+      it "Is a root branch" do
+        expect(positive_sub_branch.root?).to be       false
```
description says root branch, test says no :man_shrugging:

PragTob commented on spec/source_file_branch_spec.rb:
```diff
@@ -0,0 +1,151 @@
+        expect(positive_sub_branch.badge).to eq "+"
+      end
+
+      it "Has badge - of negative branch" do
```
this is a bit confusing to read as the tests directly contradict each other (has negative badge, and has positive badge) - I think it might be improved by introducing `describes` that focus on root/positive/negative branch respectively. What do you think?

PragTob commented on spec/source_file_branch_spec.rb:
```diff
@@ -0,0 +1,151 @@
+        expect(negative_sub_branch.sub_branch?).to be true
+      end
+
+      it "Has badge + of positive branch" do
```
I think usually these start lower case :)

PragTob commented:
Looking at #700 it seems to me that most of the relevant changes are in spec_files and that's basically an indentation change I think/hope that your merge conflicts against #700 should be relatively easy to handle. I'm tempted to merge it first, which would be bad for you but I think I can do a PR on your branch to resolve them. What do you think @som4ik ?

som4ik commented on README.md:
````diff
@@ -589,6 +589,15 @@ You can also entirely refuse dropping coverage between test runs:
+
+```ruby
+SimpleCov.start do
+  use_branchable_report true
````
It's a perfect idea. I will apply it and add validation on it.

som4ik commented on lib/simplecov.rb:
```diff
@@ -254,6 +236,79 @@ def result_exit_status(result, covered_percent)
+    #
+    def start_coverage_measurment
+      if branchable_report
+        Coverage.start(:all)
```
Agree this feature is not documented, you can have a look at this article:
https://blog.bigbinary.com/2018/04/11/ruby-2-5-supports-measuring-branch-and-method-coverages.html
A comment on the other PR: https://github.com/colszowka/simplecov/pull/692#discussion_r281836176

PragTob replied:
Yeah I read both of those, but having no official docs sucks. Might wanna PR against ruby. We could include at least one of those links at docs for our future selves :)

som4ik commented:
Yep, I agree,  #700 should be merged first.

som4ik commented on lib/simplecov.rb:
```diff
@@ -254,6 +236,79 @@ def result_exit_status(result, covered_percent)
+      if branchable_report
+        Coverage.start(:all)
+      else
+        Coverage.start
```
Yep, right!

PragTob replied:
Self mental note, I think we can't since not all ruby versions support that argument

som4ik commented on lib/simplecov/branches_per_file.rb:
```diff
@@ -0,0 +1,27 @@
+  # Doing something similar to https://ruby-doc.org/stdlib-2.5.3/libdoc/coverage/rdoc/Coverage.html can throw an error.
+  # Errors can be related to Constants, modules are included/extended inside the file but not yet loaded or even exist.
+  # It even can be a programmer's error.
+  module BranchesPerFile
```
Frankly, somehow I agree with @klyonrad, but on the other hand, this mechanism will help us to determine the branches in the missed files, which I think is nice to have. 
 After the @klyonrad comment, I added a rescue to print out the error message that problematic file raises, don't you agree with this approach? it would make these modifications not harmful at all

PragTob replied:
Hey - I'm not as worried about the errors (which we can and are catching), but more what random ruby scripts lying around might do. Imagine the following scenario:

1. I wrote ruby script to that sshs into production and runs an expensive operation there, I was carless and included all the credentials. I just dropped it somewhere in my app folder
2. I run my tests
3. The file naturally isn't covered but gets picked up through track files or what not
4. this file requires the file which causes all the ruby code to execute :scream: 
5. Suddenly production is on fire because of that expensive operation and nobody has any idea why :fire: 

I admit it's a worst case scenario. I'm certainly known to have little ruby scripts lying around that do something. They should never be executed as a side effect of a test suite run and to the best of my understanding this makes that possible. We should avoid that at all costs imo.

klyonrad replied:
Not like the ultra worst case scenario that you are painting here but at my previous job we literally had a file that **upon loading** was reading stuff from the database, storing into some kind of module/class variable and suddenly the information was shared between the different tests.

The lesson is: `require` does things 😉

som4ik commented on lib/simplecov/combiners/base_combiner.rb:
```diff
@@ -0,0 +1,32 @@
+    private
+
+      def empty_coverage?
+        first_coverage && second_coverage
```
Ops.. I am reading it now, it's a bit confusing agree :partly_sunny: 
in combiners:
`return existed_coverage unless empty_coverage?`
when `empty_coverage?` expected to return nil if any of the coverage values is nil :open_mouth: 
I would suggest changing it to: 
```
      def both_coverages_present?
        !!(first_coverage && second_coverage)
      end
```
What do you think?

PragTob replied:
Sounds good :ok_hand: 

I wouldn't even need the `!!` but that's arguably a matter of taste :)

som4ik commented on lib/simplecov/combiners/base_combiner.rb:
```diff
@@ -0,0 +1,32 @@
+        first_coverage && second_coverage
+      end
+
+      def existed_coverage
```
Yes, this method should return the coverage that exists, do you think that the method name is confusing here?

PragTob replied:
Yeah I think it should be `existing_coverage` as `existed` implies past tense (aka it is gone now :D)

som4ik commented on lib/simplecov/combiners/branches_combiner.rb:
```diff
@@ -0,0 +1,42 @@
+      #
+      def combine_branches
+        combined_result = first_coverage.clone
+        first_coverage.each do |(condition, branches_inside)|
```
Yep, agree :+1:

som4ik commented on lib/simplecov/combiners/files_combiner.rb:
```diff
@@ -0,0 +1,72 @@
+
+        combine_lines_coverage
+
+        combine_branches_coverage
```
I really prefer to have them in separate methods but on the other hand, declaring the combiners without methods will make easier to read. I will change it :wink:

som4ik commented on lib/simplecov/combiners/files_combiner.rb:
```diff
@@ -0,0 +1,72 @@
+      #
+      # @return [Hash]
+      #
+      def combine_lines_coverage
```
Yes, but according to the comment above I will remove this method

PragTob commented:
I'll get straight on that. Thanks for checking out the review and chiming in on this! :pray:

som4ik commented:
You are welcome, I have rolled over most of the comments, will continue reviewing them this evening or next two days. I will push the updates to this branch according to the code review notices ASAP.

PragTob commented:
@som4ik thank you! Don't stress out about it though, you see how long it took me to get to this :sweat_smile: We're all doing this in our free time, take time if you need it. Also feel free to tell me if you disagree with my points :)

PragTob commented:
Also, all the drop rubies support of ruby xx PRs have been merged so this is ready to be rebased/have merge conflicts resolved (hopefully). As expected the merge conflicts are in not tooo many files.

If tests fail on JRuby on travis, travis has wird problems installing rvm these days it seems :man_shrugging:

stepozer commented:
@som4ik also on our project we found that this PR working incorrect for parallel tests (gem parallel_tests)... Right now we trying to find where an issue, will let you know.

stepozer commented:
Also in HTML report I don't see new markers for branch coverage... I set up

```
SimpleCov.start('rails') { use_branchable_report true }
```

what I am doing wrong?

jeremyevans commented:
@stepozer You probably need a version of simplecov-html installed with this pull request applied: https://github.com/colszowka/simplecov-html/pull/64

tycooon commented:
It looks like merging of method coverage doesn't work in this version?

tycooon commented on lib/simplecov/source_file.rb:
```diff
@@ -149,46 +246,27 @@ def relevant_lines
-    def missed_lines
-      @missed_lines ||= lines.select(&:missed?)
-    end
+    def branches_coverage_precent
```
Shouldn't it be `branches_coverage_percent `?

som4ik replied:
Updated! Just in case, some of these methods are used in the default formatter and needs to be changed in both places.

tycooon commented on lib/simplecov/supports/source_file_support.rb:
```diff
@@ -0,0 +1,221 @@
+      # @return [Array]
+      #
+      def covered_branches
+        @coverd_branches ||= root_branches.flat_map do |root_branch|
```
I guess it's better to name this ivar `@covered_branches` so that it matches the method name.

0exp commented:
When? :)

PragTob commented:
For everyone waiting for this, thanks to a fund from [rubytogether](https://rubytogether.org/) I've got quite some good dedicated time to spend on simplecov during the next 3 months. This is among the top of my list to adjust, merge it and release it! :tada: 

![IMG_20180603_163802](https://user-images.githubusercontent.com/606517/70036138-733e9800-15b4-11ea-94c3-67a10d633cbb.jpg)

tycooon commented:
Hi @PragTob, nice to hear that! This branch is a nice work but unfortunately it has some problems:

- JSON serialization / deserialization of results does not work, so basically this doesn't work in case you want to merge results from different jobs
- Branch coverage doesn't consider `:nocov:` blocks
- No method coverage support (although this one might be out of scope of this PR)

I've fixed all these problems here (the fork of this PR) https://github.com/umbrellio/simplecov/commits/add-branch-and-method-coverage and we use this fork in our projects to measure all 3 types of coverage (line, branch, method) and also we have parallelized build, so we also successfully merge the results from different json-files.

All this stuff already works, but that branch is not ready for PR (need more tests and also some refactoring).

So basically we have 2 options now:
- Wait for @som4ik or someone else to fix the before mentioned problems in current PR (at least first 2 of them).
- Wait for me to finish my fork and create a new PR from that.

What do you think about that?

stepozer commented:
Hello everyone! Right now we successfully using this PR on our coder more than 4 months. We fixed a lot of bugs with parallel tests, merge results etc. We have some monkey patches that fixes all of this issues. If it interested for you - i will pass all of them into this comment. Unfortunately I don't have time create different PR for this.

```
module SimpleCov
  module Supports
    module SourceFileSupport
      def branches_collection(given_branches, root_id = nil)
        @branches_collection ||= []
        given_branches.each do |branch_args_str, value|
          parsed_value = JSON.parse(JSON.dump(value))
          branch_args = eval(branch_args_str)
          branch_args << root_id
          branch = SourceFile::Branch.new(*branch_args)

          if parsed_value.is_a?(Integer)
            branch.coverage = parsed_value
          else
            branches_collection(parsed_value, branch.id)
          end

          @branches_collection << branch
        end
        @branches_collection
      end
    end
  end
  module ResultMerger
    class << self
      def merge_results(*results)
        parsed_results = JSON.parse(JSON.dump(results.map(&:original_result))).map { |x| x.transform_values(&:symbolize_keys) }
        combined_result = SimpleCov::RunResultsCombiner.combine!(*parsed_results)
        result = SimpleCov::Result.new(combined_result)
        result.command_name = results.map(&:command_name).sort.join(', ')
        result
      end
    end
  end
  class Result
    def initialize(original_result)
      @original_result = original_result.freeze
      @files = SimpleCov::FileList.new(original_result.map do |filename, coverage|
        SimpleCov::SourceFile.new(filename, JSON.parse(JSON.dump(coverage)).symbolize_keys) if File.file?(filename)
      end.compact.sort_by(&:filename))
      filter!
    end
  end
  module Combiners
    class BaseCombiner
      attr_reader :first_coverage, :second_coverage

      def self.combine!(*args)
        new(*args).combine
      end

      def initialize(first_coverage, second_coverage)
        @first_coverage = first_coverage
        @second_coverage = second_coverage
      end

      private

      def empty_coverage?
        first_coverage && second_coverage
      end

      def existed_coverage
        first_coverage || second_coverage
      end
    end
    class FilesCombiner < BaseCombiner
      #
      #
      # @param [Hash] first_coverage
      # @param [Hash] second_coverage
      #
      def initialize(first_coverage, second_coverage)
        @combined_results ||= {}
        super
      end

      #
      # Handle combining results
      # => Check if any of the files coverage is empty or not
      # => Call lines combiner
      # => Call Branches combiner
      # Notice: this structure gives possibility to add in future methods coverage combiner
      #
      # @return [Hash]
      #
      def combine
        return existed_coverage unless empty_coverage?

        combine_lines_coverage

        combine_branches_coverage

        combined_results
      end

      #
      # Merge combined lines coverage results inside total results hash
      #
      # @return [Hash]
      #
      def combine_lines_coverage
        combined_results[:lines] = LinesCombiner.combine!(
          first_coverage[:lines],
          second_coverage[:lines]
        )
      end

      #
      # Merge combined branches coverage results inside total results hash
      #
      # @return [Hash]
      #
      def combine_branches_coverage
        combined_results[:branches] = BranchesCombiner.combine!(
          first_coverage[:branches],
          second_coverage[:branches]
        )
      end

      private

      # rubocop:disable TrivialAccessors
      # attr_reader under private for ruby < 2.3.7 raise "warning: private attribute?"
      def combined_results
        @combined_results
      end
      # rubocop:enable TrivialAccessors
    end
    class LinesCombiner < BaseCombiner
      def combine
        return existed_coverage unless empty_coverage?

        combine_lines
      end

      def combine_lines
        first_coverage.map.with_index do |first_coverage_val, index|
          second_coverage_val = second_coverage[index]
          merge_line_coverage(first_coverage_val, second_coverage_val)
        end
      end

      #
      # Return depends on value
      #
      # @param [Integer || nil] first_val
      # @param [Integer || nil] second_val
      #
      # Logic:
      #
      # => nil + 0 = nil
      # => nil + nil = nil
      # => int + int = int
      # @return [Integer || nil]
      #
      def merge_line_coverage(first_val, second_val)
        sum = first_val.to_i + second_val.to_i

        if sum.zero? && (first_val.nil? || second_val.nil?)
          nil
        else
          sum
        end
      end
    end
    class BranchesCombiner < BaseCombiner
      def combine_branches
        combined_result = first_coverage.clone
        first_coverage.each do |(condition, branches_inside)|
          branches_inside.each do |(branch_key, branch_coverage_value)|
            compared_branch_coverage = second_coverage.dig(condition, branch_key)

            combined_result[condition][branch_key] = branch_coverage_value + compared_branch_coverage.to_i
          end
        end

        combined_result
      end
    end
  end

  module BranchesPerFile
    def self.start(_source_file_path)
      {}
    end
  end
end

```

PragTob commented:
@tycooon Hi, thanks for the input! :tada: 

My thinking was that I'd put the work in to fix up this branch. As you can see with my longer breaks in maintaining simplecov "properly" I know how hard it is to get the dedicated time to do this and I'm happy to have it. Besides, I absolutely expect there to be problems with branch coverage etc. so having reworked it to get in would also be helpful for me because it means I know the code on another level :grin: 

@tycooon is your fork a fork of this PR? I was inclined to start here, as I already read and understood (or thought so...) the code once. If you have implemented everything already that might be a better starting point for me... or I could have a look at it to see what else was changed :)

PragTob commented:
@stepozer thank you that is really helpful! :tada: :rocket: :green_heart: 

![IMG_20180622_182841-ANIMATION](https://user-images.githubusercontent.com/606517/70041640-d254da80-15bd-11ea-94f7-125a70be0e33.gif)

som4ik commented:
@tycooon  Sorry for the late delay! soon I will push an update for this PR! 
@stepozer  Thank you :+1:  very useful!

PragTob commented:
@som4ik don't feel pressured to update the PR, as I said I was planning on updating it myself. I'll leave a comment here before I do though (and have to do it on another branch either way). If you get to upgrading it it's of course also welcome - it's a community effort after all :)

tycooon commented:
> @tycooon is your fork a fork of this PR? 
Yep, although I changed some stuff :)

som4ik commented:
@PragTob I have done: 
- Squash for all my prev commits
- Resolved the conflicts with the latest update master branch 
- Removed the issue related to `branches_per_file.rb` (removed require totally)
- Fixed code review notices
- Added suggested the ability to handle parallel tests by @stepozer  
(@stepozer I tested it with `parallel_tests` gem it seems everything ok, if you could verify it I will appreciate it :) ) 

@tycooon I reviewed the changes/upgrades that you have done, I like it, good job! :+1:  
Do you have any updates for the default `simplecov-html` formatter that covers Methods coverage reports? :) 
As you already added the methods combiner, I didn't change the name of the config input `use_branchable_report` or changed it's functionality as we discussed [before](https://github.com/colszowka/simplecov/pull/694#discussion_r298918548), I thought it would be nice to be done in your updates or PR, what do you think? 
![cute-bunny-gif](https://user-images.githubusercontent.com/7699543/70100545-1d650100-1643-11ea-81e6-efeaf6cb3667.gif)

stepozer commented:
@som4ik I will try update code on our project. I'll let you know how your new code works

tycooon commented:
> @tycooon I reviewed the changes/upgrades that you have done, I like it, good job! 👍
> Do you have any updates for the default `simplecov-html` formatter that covers Methods coverage reports? :)

Nope, was going to add that later. 

> As you already added the methods combiner, I didn't change the name of the config input `use_branchable_report` or changed it's functionality as we discussed [before](https://github.com/colszowka/simplecov/pull/694#discussion_r298918548), I thought it would be nice to be done in your updates or PR, what do you think?

You are proposing that I add method coverage support in a later PR, am I correct? In that case we will definitely need to rename that method, so maybe it's better to rename it now, so that we don't have to care about the backward compatibility later?

PragTob commented:
@som4ik Thank you so much for all this :tada: I'm currently runnign a bunch of errands, but will re-review some time later today.

@tycooon I think getting this merged is a bit more important (tiny bit) than the right configuration option and method coverage or what not to boot with. I'm fine merging this now but waiting to cut a release until we changed the naming to be better. I still like `coverage` or `coverage_type` with a symbol, I was gonna look up in the literature how they usually differentiate the different types of coverage/call them.

@som4ik your recent email address doesn't seem to be connected to your github account, just fyi if you care about these kinds of things.

Thanks everyone!

![IMG_20190215_122642_Bokeh](https://user-images.githubusercontent.com/606517/70133401-4e1a5a00-1686-11ea-852c-8c3db9a7be0a.jpg)

PragTob commented on lib/simplecov/source_file.rb:
```diff
@@ -72,13 +73,105 @@ def status
+      #
+      def positive?
+        return true if type == :when
+        1 + root_id.to_i == id
```
Highlighting this once more cos I still got no idea what's happening here :grin:

som4ik replied:
For `if, else` conditions:
coverage returns matrices ex: [:if, 0,..] => {[:then, 1,..], [:else, 2,..]},
where positive branch always has Id equals to root branch Id incremented by 1.

som4ik replied:
I will add more comments to this method

PragTob replied:
Guessed as much but wanted confirmation. Thank you! I guess it also works for `unless` which then is the question should that still be called positive but I think so. Thank you!

PragTob commented on lib/simplecov/source_file.rb:
```diff
@@ -149,46 +246,27 @@ def relevant_lines
-      @missed_lines ||= lines.select(&:missed?)
-    end
+    def branches_coverage_percent
+      return 100.0 if no_branches? && no_lines?
```
still not sure why we need to check `no_lines?` :)

som4ik replied:
Yep, It should be removed, I will do it!

stepozer commented:
Good job. Will nice if you release this and I will update gem and check.

som4ik commented:
@PragTob  I added more comments in some places and also done small changes that give significant upgrade for the report matrics.
BEFORE: 
![befor-final](https://user-images.githubusercontent.com/7699543/70222881-b71fd180-175b-11ea-95ce-0efe14667668.png)
AFTER:
![after-final](https://user-images.githubusercontent.com/7699543/70222903-bf780c80-175b-11ea-8b68-1f1b4a783ade.png)
Hope CI will pass :pray: :fist_raised: :wink:

PragTob commented:
@som4ik that looks nice! I think we could work on the presentation a bit but in a separate step :)

That script looks like it gives a nice overview of branch coverage cases - I think it might be a good to include it in the repo in an integration test - I could type it all from the image but if you have, I'd appreciate it :)

PragTob commented:
Ok, let's do this :rocket:

som4ik commented:
```
x = 1
x.eql?(4) ? "4" : x

puts x unless x.eql?(5)

unless x == 5
  puts "Ola.."
end

unless x != 5
  puts "Ola.."
end

unless x != 5
  puts "Ola.."
else
  puts "text"
end

puts x if x.eql?(5)
if x != 5
  puts "Ola.."
end

if x == 5
  puts "Ola.."
end

if x != 5
  puts "Ola.."
else
  puts "text"
end
x = 4
if x == 1
  puts x

  puts x
  puts x
elsif x == 4
  puts x
  puts x
  puts x
  puts x
elsif x == 14
  puts x

  puts x
else
  puts x
end

if x == 1
  puts x
elsif x == 4
  puts x
elsif x == 14
  puts x
else
  puts x
end

stepozer commented:
We updated to new version, seems to be working. Will let you now if something will be wrong.

PragTob commented:
Remaining work is tracked in #781

PragTob commented:
@stepozer thank you very much! :tada:

hedgesky commented:
So cool this got merged finally, thanks a lot @som4ik, @PragTob!
